datalad/distribution/dataset.py

```diff
@@ -255,12 +261,20 @@ def config(self):
         -------
         ConfigManager
         """
-        if self._cfg is None:
-            if self.repo is None:
-                # associate with this dataset and read the entire config hierarchy
-                self._cfg = ConfigManager(dataset=self, dataset_only=False)
-            else:
-                self._cfg = self.repo.config
+
+        if self.repo is None:
+            # if there's no repo (yet or anymore), we can't read/write config at
+            # dataset level, but only at user/system level
+            # However, if this was the case before as well, we don't want a new
+            # instance of ConfigManager
+            if self._cfg_bound in (True, None):
+                self._cfg = ConfigManager(dataset=None, dataset_only=False)
+                self._cfg_bound = False
+
+        else:
+            self._cfg = self.repo.config
+            self._cfg_bound = True
```

Comment:
Yes, but it's just a pointer assignment. The only thing really not lazy here is `repo` (and yes -that's a performance issue). But this is not to fix within `config`. Here we need to ensure we are pointing to the correct `ConfigManager` instance. If `repo` changes, we get an entirely inconsistent `Dataset.config` otherwise.